PyDSS/pyControllers/Controllers/MotorStall.py

```diff
@@ -5,94 +5,186 @@
+                    self._ControlledElm.SetParameter('kw', self.kw_rated * self.__Settings['Pfault'] )#* abs(p_stall_pu))
+                    self._ControlledElm.SetParameter('kvar', self.kvar_rated * self.__Settings['Qfault']) #*  abs(p_stall_pu))
+                    self.model_mode = self._ControlledElm.SetParameter('model', 2)
+                      
+            if self.voltage_pu < V_stall_adj and not self.stall:
+                self.stall_time_start = self.__dssSolver.GetTotalSeconds()
+                self.stall = True
+            
+            u = 1 if self.model_mode == 2 else 0
+            
+            self.T = np.array([self.T[-1], self.t])
+            self.U = np.array([self.U[-1], u])
+            tout, yout, xout = signal.lsim2(self.H, self.U, self.T, self.X)
+            self.X = xout[-1]
+            theeta = yout[-1]
+            
+            if theeta < self.__Settings['Tth1t']:
+                Kth = 1
+            elif theeta > self.__Settings['Tth2t']:
+                Kth = 0
+            else:
+                m = 1 / (self.__Settings['Tth1t'] - self.__Settings['Tth2t'])
+                c = - m * self.__Settings['Tth2t']
+                Kth = m * theeta + c
+            
+            if self.model_mode == 2:
+                p_stall = self.kw_rated * self.__Settings['Pfault'] #* abs(p_stall_pu)
+                q_stall = self.kvar_rated * self.__Settings['Qfault'] #* abs(p_stall_pu)
+                
+                self._ControlledElm.SetParameter('kw', p_stall * Kth)
+                self._ControlledElm.SetParameter('kvar', q_stall * Kth)
+
+            if Kth==0 and self.voltage_pu > self.__Settings['Vrstrt']:
+                self.v = 1
+                
+            self.rT = np.array([self.rT[-1], self.t])
+            self.rU = np.array([self.rU[-1], self.v])
+            _, yout_r, xout_r = signal.lsim2(self.R, self.rU, self.rT, self.rX)
+            self.rX = xout_r[-1]
+            reconnect = yout_r[-1]
+
+            if reconnect > 0:
+                self.model_mode = self._ControlledElm.SetParameter('model', self.mode)
+                self._ControlledElm.SetParameter('kw', self.kw_rated * reconnect)
+                self._ControlledElm.SetParameter('kvar', self.kvar_rated * reconnect)
+                self.stall = False
+                
+            ######################################### TESTING CODE ###################################################
+            self.f.write(f"{self.current_pu},{self.voltage_pu},{self.p},{self.q},{u},{tout[-1]},{yout[-1]},{Kth},{CompLF},{CompPF},{V_break_adj},{V_stall_adj}\n")
+            ##########################################################################################################
+        self.model_mode_old = self.model_mode
         return 0
 
+    def __del__(self):
+        
+        ######################################### TESTING CODE ###################################################
+        try:
```

Comment:
This could certainly fail if there are other exceptions and GC runs much later. It would be better if the code could be run inside a context manager - if that is even possible (I have an idea on how it could be done if you want to discuss it). Another type of error is one in which an exception gets raised before `self.f` gets assigned. You could check that here if you want.

Reply:
removed